libs/stokesian_dynamics/device_matrix.hpp

```diff
@@ -0,0 +1,658 @@
+// Copyright (c) 2018 Henri Menke. All rights reserved.
+
+#include <iomanip>
+#include <limits>
+#include <type_traits>
+
+#include "thrust_wrapper.hpp"
+
+#ifdef __CUDACC__
+#include <cublas_v2.h>
+#include <cusolverDn.h>
+#endif
+
+#ifdef __CUDACC__
+#define DEVICE_FUNC __host__ __device__
+#else
+#define DEVICE_FUNC
+#endif
+
+#ifdef __GNUC__
+#define MAYBE_UNUSED __attribute__((unused))
+#else
+#define MAYBE_UNUSED
+#endif
+
+namespace policy {
+
+struct host {
+    template <typename T>
+    using vector = thrust_wrapper::host_vector<T>;
+    // the remove_const is necessary, because things cannot be static and const
+    // at the same time
+    static std::remove_const<decltype(thrust_wrapper::host)>::type par()
+      { return thrust_wrapper::host; }
+};
+
+struct device {
+    template <typename T>
+    using vector = thrust_wrapper::device_vector<T>;
+    static std::remove_const<decltype(thrust_wrapper::device)>::type par()
+      { return thrust_wrapper::device; }
+};
+
+template <typename...>
+using void_t = void;
+
+template <typename Policy, typename = void>
+struct is_policy : std::false_type {};
+
+template <typename Policy>
+struct is_policy<Policy, void_t<typename Policy::template vector<class T>,
+                                decltype(Policy::par())>> : std::true_type {};
+
+} // namespace policy
+
+DEVICE_FUNC inline thrust_wrapper::tuple<std::size_t, std::size_t>
+unravel_index(std::size_t index, std::size_t lda) {
+    std::size_t i, j;
+    i = index % lda;
+    index /= lda;
+    j = index;
+    return thrust_wrapper::make_tuple(i, j);
```

Comment:
Why is there this `thrust_wrapper` garbage in my code?  I have told you before that this code requires Thrust.  By replacing this you have fucked up the main advantage of my architecture which is that you can generate parallel code on the GPU *and* the CPU from a single source by cleverly using the Thrust backends.  What is the problem with adding Thrust as a Git submodule?

Reply:
I can't tell about adding Thrust as a git submodule, but as long as Thrust is available, the advantage you mentioned stays the same (i.e. parallel execution on CPU). If Thrust isn't available, it can be compiled nonetheless, using std-library routines.

Reply:
The clever architecture hasn't changed. If Thrust is available, it is used as before. If it's not, GPU code is not compiled and everything runs on the GPU. This should essentially be a `sed s/thrust::/thrust_wrapper::/g` and some changed header inclusions compared to your code.

Reply:
> Why is there this `thrust_wrapper` garbage in my code? I have told you before that this > >code requires Thrust. By replacing this you have fucked up the main advantage of my >architecture which is that you can generate parallel code on the GPU _and_ the CPU from a >single source by cleverly using the Thrust backends. What is the problem with adding Thrust >as a Git submodule?

Henri, independently of the technical issue under debate, I find the tone of your comment inappropriate. I perceive it as hurtful with regards to the people working on this PR and I think it is harmful for the project overall.

Reply:
I apologize for my strong language and I have edited my previous comment.  However, I am still a bit unhappy about the direction this is taking.  We should have a discussion about this.

Reply:
We could definitely do that.